### PR TITLE
one tutorial to rule them all

### PR DESCRIPTION
**Pull request by joelgrus**
I started with the PyTorch sequence tagging tutorial: https://pytorch.org/tutorials/beginner/nlp/sequence_models_tutorial.html

My goal was to make a tutorial that should be "the only tutorial you need" for like 95% of our users. You tell me if this accomplishes that.

### CODE REVIEW

File: tutorials/tagger/pos_allennlp_bare.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+from allennlp.data.fields import TextField, SequenceLabelField
+from allennlp.data.token_indexers import SingleIdTokenIndexer
+from allennlp.data.tokenizers import Token
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.models import Model
+from allennlp.modules.feedforward import FeedForward
+from allennlp.modules.text_field_embedders import TextFieldEmbedder, BasicTextFieldEmbedder
+from allennlp.modules.token_embedders import Embedding
+from allennlp.modules.seq2seq_encoders import Seq2SeqEncoder, PytorchSeq2SeqWrapper
+from allennlp.nn.util import get_text_field_mask, sequence_cross_entropy_with_logits
```
Comment (joelgrus) [inline tutorials/tagger/pos_allennlp_bare.py]: that's a lot of imports!

Comment (matt-gardner) [reply tutorials/tagger/pos_allennlp_bare.py]: You get around this in the plain pytorch version by importing `nn` and doing `nn.X`.  You could do the same here, if you wanted to.

**Comment by schmmd** [comment]
Closing in favor of https://github.com/allenai/allennlp-website/pull/73.

**Comment by joelgrus** [comment]
I need to still merge this into the main tutorials codebase (along with the code required to transform it into the right format)

File: tutorials/tagger/basic_allennlp_annotated.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                  patience=10,
+                  num_epochs=1000)
+
+#### When we launch it it will print a progress bar for each epoch that includes both the "loss" and the "accuracy" metric. If our model is good, the loss should go down and the accuracy up as we train.
+trainer.train()
+
+#### As in the original PyTorch tutorial, we'd like to look at the predictions our model generates. AllenNLP contains a <code>Predictor</code> abstraction that takes inputs, converts them to instances, feeds them through your model, and returns JSON-serializable results. Often you'd need to implement your own Predictor, but AllenNLP already has a <code>SentenceTaggerPredictor</code> that works perfectly here, so we can use it. It requires our model (for making predictions) and a dataset reader (for creating instances).
+predictor = SentenceTaggerPredictor(model, dataset_reader=reader)
+#### It has a <code>predict</code> method that just needs a sentence and returns (a JSON-serializable version of) the output dict from forward.  Here "tag_scores" will be a (5, 3) array of logits, corresponding to the 3 possible tag for each of the 5 words.
```
Comment (matt-gardner) [inline tutorials/tagger/basic_allennlp_annotated.py]: 3 possible tags

Comment (matt-gardner) [reply tutorials/tagger/basic_allennlp_annotated.py]: And when I first read this I was confused at the mismatch between "tag_logits" and "tag_scores".  Also, "tag_scores" in the annotation should probably be in a <code> block.

File: tutorials/tagger/basic_allennlp_annotated.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+    #### We included an accuracy metric that gets updated each forward pass. That means we need to override a <code>get_metrics</code> method that pulls the data out of it. Behind the scenes, the <code>CategoricalAccuracy</code> metric is storing the number of predictions and the number of correct predictions, updating those counts during each call to forward. Each call to get_metric returns the calculated accuracy and (optionally) resets the counts, which is what allows us to track accuracy anew for each epoch.
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        return {"accuracy": self.accuracy.get_metric(reset)}
+
+
+#### Now that we've implemented a <code>DatasetReader</code> and <code>Model</code>, we're ready to train. We first need an instance of our dataset reader.
+reader = PosDatasetReader()
+#### Which we can use to read in the training data and validation data.
+train_dataset = reader.read('tutorials/tagger/training.txt')
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tutorials/tagger/basic_allennlp_annotated.py]: And if you make it so that you can just copy and paste this whole thing and run it, I might even mention that up at the top.

**Comment by joelgrus** [comment]
ok, I think I addressed all of your comments, except that I didn't change the validation metric because on a dataset this small the training spends a long time with accuracy not improving and so I would have had to set the patience to a really huge number to get it to train

File: tutorials/tagger/basic_allennlp.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        self.hidden2tag = torch.nn.Linear(in_features=encoder.get_output_dim(),
+                                          out_features=vocab.get_vocab_size('labels'))
+        #### The last thing to notice is that we also instantiate a <code>CategoricalAccuracy</code> metric, which we'll use to track accuracy during each training and validation epoch.
+        self.accuracy = CategoricalAccuracy()
+
+    #### Next we need to implement <code>forward</code>, which is where the actual computation happens. Each <code>Instance</code> in your dataset will get (batched with other instances and) fed into <code>forward</code>. The <code>forward</code> method expects dicts of tensors as input, and it expects their names to be the names of the fields in your <code>Instance</code>. In this case we have a sentence field and (possibly) a labels field, so we'll construct our <code>forward</code> accordingly:
+    def forward(self,
+                sentence: Dict[str, torch.Tensor],
+                labels: Dict[str, torch.Tensor] = None) -> torch.Tensor:
```
Comment (matt-gardner) [inline tutorials/tagger/basic_allennlp.py]: This type annotation should still be a `Tensor.

File: tutorials/tagger/basic_allennlp.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+        if tags:
+            label_field = SequenceLabelField(labels=tags, sequence_field=sentence_field)
+            fields["labels"] = label_field
+
+        return Instance(fields)
+
+    #### The other piece we have to implement is <code>_read</code>, which takes a filename and produces a stream of <code>Instance</code>s. Most of the work has already been done in <code>text_to_instance</code>.
+    def _read(self, file_path: str) -> Iterator[Instance]:
+        with open(file_path) as f:
```
Comment (matt-gardner) [inline tutorials/tagger/basic_allennlp.py]: Did you actually try copying and pasting this code and running it?  I think you need a `cached_path` here.

Comment (joelgrus) [reply tutorials/tagger/basic_allennlp.py]: I didn't run it with the URLs because the files aren't on GitHub until I merge this PR 😢

File: tutorials/tagger/basic_allennlp.py (hunk 4 above)
Comment (joelgrus) [reply tutorials/tagger/basic_allennlp.py]: fixed

File: tutorials/tagger/basic_allennlp.py (hunk 5 above)
Comment (joelgrus) [reply tutorials/tagger/basic_allennlp.py]: fixed
